### PR TITLE
RFC: Add ZeroMQ as a transport layer for xpcc.

### PR DESCRIPTION
**Pull request by strongly-typed**
I opened this PR for discussion.

There are probably some to many fixes and rebases necessary until this is added to mainline. 

But I still think it's worth to use ZeroMQ at RCA after it was successfully demonstrated to control a robot from the internal JavaGUI by ZeroMQ. This PR should bring this effort forward.

### CODE REVIEW

**Comment by ekiwi** [comment]
> You are right. Now fixed

Thank you.

One last question: Why do we need to support zeromq versions < 4.1.1?
If there is a valid use case that still uses an old zeromq version I am for including it, we just have to make sure to keep the maintenance burden small and one way of doing this is to only support the latest version of a library.

**Comment by strongly-typed** [comment]
Thank for your feedback!

> Why do we need to support zeromq versions < 4.1.1?

I think that on the RPi mounted on the big bot only libzmqpp3 was available. On OSX (with homebrew) I started to develop with libzmqpp4 right away.

> is to only support the latest version of a library.

New is always better? Hmm :-) 
I think of a project at work where a competent project partner chose zeromq3. I will ask him for his reasons.

**Comment by ekiwi** [comment]
> New is always better? Hmm :-)

of course :D, as long as I am not being paid I will only support the most recent version of any open source project that I am involved in, and I think that other developers will feel similarly.

> I think that on the RPi mounted on the big bot only libzmqpp3 was available

OK. This is a very important reason!

**Comment by salkinium** [comment]
Could you please rebase this? The diff is a bit noisy, especially when reviewing commit by commit.

**Comment by strongly-typed** [comment]
I will redo all commits and separate zeroMQ stuff from non-zeroMQ stuff.
Rebasing is a bit painful because git does not remember how to solve merge conflicts.

**Comment by strongly-typed** [comment]
So, some cleanup done. Building the example will probably fail because of #178.

**Comment by daniel-k** [comment]
>  Rebasing is a bit painful because git does not remember how to solve merge conflicts.

Maybe git `rerere` [1] is your friend.

[1] https://git-scm.com/blog/2010/03/08/rerere.html

**Comment by strongly-typed** [comment]
[Travis CI is soooo outdated.](http://packages.ubuntu.com/search?keywords=libzmqpp-dev)

**Comment by strongly-typed** [comment]
Travis CI standard is too old (no libzmqpp), Travis CI trusty beta is buggy [#6541](https://github.com/travis-ci/travis-ci/issues/6541). So no Travis CI yet.

**Comment by salkinium** [comment]
An alternative is Circle CI, but they also only have 14.04 running 😞.

**Comment by ekiwi** [comment]
There is a gitlab CI running on the latest Arch Linux at RCA. That should be up to date.
@daniel-k @rleh how difficult would it be to make build results public?

**Comment by strongly-typed** [comment]
So hopefully this compiles on Travis CI with Trusty.

**Comment by salkinium** [comment]
I'm very sorry, but F1 alternate functions is [funktional kaputt](https://www.youtube.com/watch?v=3LNYfAYE72E) and there is no quickfix for now (See https://github.com/roboterclubaachen/xpcc/issues/178#issuecomment-250970902). Do you have a board other than a F1 to write the STM32 examples with?

**Comment by strongly-typed** [comment]
The Blue Pill is so bread board friendly :-O
OK, I will then use an **ARDUINO** _scnr_

**Comment by salkinium** [comment]
Please allow me to push to this PR, I'd like to rebase and resolve the conflict.

File: examples/zmq/README.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# Simple ZeroMQ example with XPCC
+
+## Setup
+
+ * STM32F103 "Blue Pill" development board
```
Comment (salkinium) [inline examples/zmq/README.md]:
``` diff
- * STM32F103 "Blue Pill" development board
+ * STM32F407 Discovery development board
```

File: examples/zmq/README.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+ * CAN transceiver breakout board with SN65HVD230 
+ * CAN2USB
+
+Connect the CAN2USB to your computer and to the CAN transceiver.
+
+## STM32 XPCC publisher
+
+Flash `1_stm32` on the Blue Pill
```
Comment (salkinium) [inline examples/zmq/README.md]:
``` diff
- Flash `1_stm32` on the Blue Pill
+ Flash `1_stm32` on the STM32F407 Discovery
```

File: src/xpcc/communication/xpcc/backend/zeromq/connector.hpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+			case
+			Mode::PubPull:
+				this->socketIn.bind(endpointIn);
+				this->socketOut.bind(endpointOut);
+			break;
+		}
+	}
+
+	virtual
+	~ZeroMQConnector() {}
```
Comment (salkinium) [inline src/xpcc/communication/xpcc/backend/zeromq/connector.hpp]: Should this not disconnect/unbind the socket?

**Review by salkinium — CHANGES_REQUESTED** [review]
Would it be ok to treat the example like the Blue Pill was never involved? So to apply the F4 Discover change commit as a fixup to the stm32 example?

File: examples/zmq/README.md (hunk 2 above)
Comment (strongly-typed) [reply examples/zmq/README.md]: You really want to wipe out all evidence ;)

File: src/xpcc/communication/xpcc/backend/zeromq/connector.hpp (hunk 3 above)
Comment (strongly-typed) [reply src/xpcc/communication/xpcc/backend/zeromq/connector.hpp]: Probably the destructor will take care of it. I added `unsubscribe()` and `close()` where possible.

**Review by salkinium — APPROVED** [review]

**Comment by salkinium** [comment]
Rebased and merged manually.

**Pull request closed without merging**